src/strands/multiagent/graph.py

```diff
@@ -487,22 +537,79 @@ async def _execute_graph(self, invocation_state: dict[str, Any]) -> None:
             if not should_continue:
                 self.state.status = Status.FAILED
                 logger.debug("reason=<%s> | stopping execution", reason)
-                return  # Let the top-level exception handler deal with it
+                return
 
             current_batch = ready_nodes.copy()
             ready_nodes.clear()
 
-            # Execute current batch of ready nodes concurrently
-            tasks = [asyncio.create_task(self._execute_node(node, invocation_state)) for node in current_batch]
-
-            for task in tasks:
-                await task
+            # Execute current batch of ready nodes in parallel
+            if len(current_batch) == 1:
+                # Single node - execute directly to avoid overhead
+                async for event in self._execute_node(current_batch[0], invocation_state):
+                    yield event
+            else:
+                # Multiple nodes - execute in parallel and merge events
+                async for event in self._execute_nodes_parallel(current_batch, invocation_state):
+                    yield event
 
             # Find newly ready nodes after batch execution
-            # We add all nodes in current batch as completed batch,
-            # because a failure would throw exception and code would not make it here
             ready_nodes.extend(self._find_newly_ready_nodes(current_batch))
 
+    async def _execute_nodes_parallel(
```

Comment:
Does this follow the pattern of how tools are executed in parallel - I know we worked out some wrinkles/bugs with that, so it'd be ideal to not have to redo those